lib/matplotlib/category.py

```diff
@@ -0,0 +1,122 @@
+"""
+catch all for categorical functions
+"""
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+
+import six
+from collections import Iterable
+
+import numpy as np
+
+import matplotlib.units as units
+import matplotlib.ticker as ticker
+
+
+class StrCategoryConverter(units.ConversionInterface):
+    @staticmethod
+    def convert(value, unit, axis):
+        """Uses axis.unit_data map to encode
+        data as integers
+        """
+        vmap = dict(axis.unit_data)
+
+        if isinstance(value, Iterable):
+            vals = np.array([vmap[str(v)] for v in value])
+        else:
+            vals = vmap[value]
+        return vals
+
+    @staticmethod
+    def axisinfo(unit, axis):
+        seq, locs = zip(*axis.unit_data)
+        majloc = StrCategoryLocator(locs)
+        majfmt = StrCategoryFormatter(seq)
+        return units.AxisInfo(majloc=majloc, majfmt=majfmt)
+
+    @staticmethod
+    def default_units(data, axis):
+        # the conversion call stack is:
+        # default_units->axis_info->convert
+        axis.unit_data = map_categories(data, axis.unit_data)
+        return None
+
+
+def map_categories(data, old_map=[], sort=True):
```

Comment:
Don't put mutables in the signatures as defaults, it can lead to some fun closures.

```python
def foo(d=[]):
    d.append(len(d))
    print(d)
```
gives
```python
In [24]: foo()
[0]

In [25]: foo()
[0, 1]

In [26]: foo()
[0, 1, 2]

In [27]: foo()
[0, 1, 2, 3]

```

Either make this an empty tuple or make the default `None` and put logic in the function like

```python
if old_map is None:
    old_map = []
```